File: ruby/aa/shuffle.rb
def shuffle(arr)
  new_arr = arr.dup
  len = arr.length
  arr.each_index do |i|
    rand_idx = i + rand(len - i)
    new_arr[i], new_arr[rand_idx] = new_arr[rand_idx], new_arr[i]
  end
  new_arr
end

arr = [1,2,3,4,5]
p shuffle(arr)

File: js/aa/shuffle.js
function shuffle(arr){
  const newArr = arr.slice(0);
  for (let i = 0; i < arr.length; i++) {
    let randIdx = i + Math.floor(Math.random() * (arr.length - i));
    [ newArr[i], newArr[randIdx] ] = [ newArr[randIdx], newArr[i] ];
  }
  return newArr;
}

let arr = [1,2,3,4,5];
console.log(shuffle(arr));
